src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+
+    public ValueTask UnregisterExtensionAsync(
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        Extension? extension = null;
+        lock (_lockObject)
+        {
+            if (_extensions.TryGetValue(assemblyFolderPath, out extension))
+            {
+                if (extension.RemoveAssemblyHandlers(assemblyFileName, out var assemblyHandlers))
+                {
+                    if (assemblyHandlers is not null)
+                    {
+                        foreach (var workspaceHandler in assemblyHandlers.WorkspaceMessageHandlers.Keys)
+                        {
+                            _workspaceHandlers.Remove(workspaceHandler);
+                        }
+
+                        foreach (var documentHandler in assemblyHandlers.DocumentMessageHandlers.Keys)
+                        {
+                            _documentHandlers.Remove(documentHandler);
+                        }
+                    }
+                }
+
+                if (extension.AssemblyHandlersCount > 0)
+                {
+                    return ValueTask.CompletedTask;
+                }
+
+                _extensions.Remove(assemblyFolderPath);
+            }
+        }
+
+        extension?.AnalyzerAssemblyLoader.Dispose();
+
+        return ValueTask.CompletedTask;
+    }
+
+    public void Reset()
+    {
+        List<Extension> extensions;
+        lock (_lockObject)
+        {
+            extensions = _extensions.Values.ToList();
+            _extensions.Clear();
+            _workspaceHandlers.Clear();
+            _documentHandlers.Clear();
+        }
+
+        foreach (var extension in extensions)
+        {
+            extension.AnalyzerAssemblyLoader.Dispose();
+        }
+    }
+
+    public void Dispose()
+        => Reset();
+
+    private async ValueTask<string> HandleExtensionMessageAsync<TArgument>(
+        TArgument argument,
+        string messageName,
+        string jsonMessage,
+        Dictionary<string, IExtensionMessageHandlerWrapper<TArgument>> handlers,
+        CancellationToken cancellationToken)
+    {
+        IExtensionMessageHandlerWrapper<TArgument> handler;
+        lock (_lockObject)
+        {
+            // handlers here is either _workspaceHandlers or _documentHandlers, so it must be protected
+            // by _lockObject.
+            if (!handlers.TryGetValue(messageName, out handler!))
+            {
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+            }
+        }
+
+        // Any exception thrown in this method is left to bubble up to the extension.
```

Comment:
thanks.  these are helpful comments.

that said, we might want to consider other approaches.  like if an extension throws, should we disable it.  That's the roslyn model for our own extensions (Analzyers/Generators/etc.).  We are very intolerant and basically go: once you throw, you're done.

Reply:
I don't have a problem implementing that.
I still want the first exception to bubble up to the extension though, otherwise the extender would not have any visibility.

Reply:
I went with unregistering all handlers for that assembly in case of exception, since the assembly is our unit of registration/unregistration.

Reply:
> I still want the first exception to bubble up to the extension though, otherwise the extender would not have any visibility.

That works for me.  I'm just hesitant about InvalidOp.  Perhaps ArgumentException would be better as we are throwing because of a problem with an argument they passed us.

Reply:
I'm ok with changing the type of the exception. I'll do that tomorrow